R/raw_cf.R

```diff
@@ -0,0 +1,550 @@
+#' @title Container for raw correlation functions
+#'
+#' @description
+#' This function \code{raw_cf()} creates containers for raw correlation functions
+#' of class \code{raw_cf}. This class is particularly designed to deal with
+#' complex and matrix-valued correlation functions emerging in statistical
+#' mechanics and quantum field theory simulations. 
+#' Arithmetic operations are defined for this class in
+#' several ways, as well as concatenation and \link{is.raw_cf} and \link{as.raw_cf}.
+#'
+#' @family raw_cf constructors
+#'
+#' @export
+raw_cf <- function () {
+  cf <- list()
+  class(cf) <- append(class(cf), 'raw_cf')
+  return (cf)
+}
+
+#' @title RAW_CF metadata mixin constructor
+#'
+#' @param nrObs Integer, number of different observables assembled in the data field of this container. 
+#' @param Time Integer, full time extent.
+#' @param nts Integer, number of time separations actually stored in the data field.
+#' @param nrStypes Integer, number of smearing types.
+#' @param dim Integer vector of "internal" dimensions for matrix-valued correlation functions. 
+#'            For a scalar correlation, this should be specified as \code{c(1,1)}.
+#'
+#' @family raw_cf constructors
+#'
+#' @export
+raw_cf_meta <- function (cf = raw_cf(), nrObs = 1, Time = NA, nrStypes = 1, dim=c(1,1), nts = Time ) {
+  stopifnot(inherits(cf, 'raw_cf'))
+  
+  cf$nrObs <- nrObs
+  cf$Time <- Time
+  cf$nts <- nts
+  cf$nrStypes <- nrStypes
+  cf$dim <- dim
+
+  class(cf) <- append(class(cf), 'raw_cf_meta')
+  return (cf)
+}
+
+#' @title Original data RAW_CF mixin constructor
+#'
+#' @param cf `raw_cf` object to extend.
+#' @param data Numeric or complex array, original data for all observables and measurements. 
+#'             This should have dimensions c(Nmeas,cf$Time*cf$nrObs*cf$nrStypes,cf$dim).
+#'             Having the internal dimensions innermost is not as efficient, but it allows
+#'             different transformations to be applied to different observables in the same
+#'             container more easily.
+#'
+#' @family raw_cf constructors
+#'
+#' @export
+raw_cf_data <- function (cf, data) {
+  stopifnot(inherits(cf, 'raw_cf'))
+  stopifnot(inherits(cf, 'raw_cf_meta'))
+  dims <- dim(data)
+  if( any( cf$dim != dims[3:length(dims)] ) ){
+    stop("Dimensions of 'data' and 'raw_cf$dim' container meta-data must match")
+  } 
+
+  # raw correlation functions are expected to be complex-valued 
+  if( mode(data) != "complex" ){
+    cf$data <- as.complex(data)
+  } else {
+    cf$data <- data
+  }
+
+  class(cf) <- append(class(cf), 'raw_cf_data')
+  return (cf)
+}
+
+#' @title Add Gamma-method error analysis to existing `raw_cf` container
+#' 
+#' @param cf Correlation function container of class 'raw_cf'
+#' @family raw_cf constructors
+#' 
+#' @export
+raw_cf_uwerr <- function(cf){
+  stopifnot(inherits(cf, 'raw_cf_data'))
+  stopifnot(inherits(cf, 'raw_cf_meta'))
+  cf$uwerr <- uwerr.raw_cf(cf)
```

Comment:
I do not quite understand why this has to be part of the `raw_cf` class. To me it seems that this is something that is computed *from* a `raw_cf`. The `uwerr` object might have a reference to the original `raw_cf` if one wants to this “audit trail”. I fear that, just as with the `cf` class, it becomes very broad and large.

Reply:
It doesn't have to be, but if one wants to have it in there, one can also attach it and use it when necessary.

Reply:
You can make the same argument about bootstrap or jacknife samples...

Reply:
The reason one might have for attaching it is how expensive this can become for large matrix-valued correlation funtions.

Reply:
I also do not like the way that bootstrap samples are handled in `cf`. This `cf_boot` is a step forward, but the `cf` class *is* too many different things, still. I just realized that what I actually have in mind would be that `cf_orig` is a separate class and a `cf_orig` *has* a `cf_orig` instance. Then we have `obj$meta$…` for the metadata and `obj$cf` for the original data. The `cf_boot` would then *have* a `cf_meta` instance and *have* a `tsboot` object.

Regarding expensive computations: I would think that if something is expensive to compute one should then keep track of that (separate) variable.

Perhaps we have different use cases for the library in mind. With the parameter-value-framework it is easy to have various values and I prefer small focused objects to work with. Also as the analysis is done with various script files and parameters one always knows what some function has been called with, for instance I do not need the fit object to remember the fit range as they are parameters and therefore always available.

On the other hand I have seen the hadron library used in an interactive session where more or less one `cf` object has been massaged to be various things through the functions.

Reply:
I currently have this structure in my head:

- All the `cf_` mixins become mutually exclusive classes.
- The `cf_orig` object has a member `meta` of class `cf_meta` and a member `cf` with the original data.
- In `bootstrap.cf` an instance of `cf_orig` is required. The return value is an instance of class `cf_boot` which has the member `meta` copied from input as one unit. There is no original data copied into `cf_boot`.
- This would also mean that a function like `invalidate.samples` becomes obsolete as there is no “going back”, one just uses the object with the original data.
- Stuff like the principal correlator just have a member of type `cf_boot` and the `gevp_reference_time` and `correlator_id` directly.
- Forwarding things like metadata becomes one instruction because it is just one structured element to copy.

So the thing with the current mixin approach that I am not happy with is that I believe that we actually have a composition hierarchy but do that in a flat namespace of a single object with multiple classes.

Reply:
> So the thing with the current mixin approach that I am not happy with is that I believe that we actually have a composition hierarchy but do that in a flat namespace of a single object with multiple classes.

Yes, I see the point. Moving away from the present setup, however, means that one will have to 

1) either construct a container externally to keep things together (when working with these things interactively), thus having to implement all the stuff that the ```cf``` class currently takes care of automatically (such as invalidating samples under transformations).
2) keep track of many objects, which is annoying interactively

I agree that inside a framework, working with small objects is often easier. However, exploratory analysis requires lots of meta-information and is ideally done interactively to make quick progress, at least that has been my experience so far. Once everything is packed up into some framework, it becomes somewhat difficult to follow what one is actually doing. Case in point is the difficulty that we faced extracting a particular parameter combination from a paramvalf container...

Reply:
I probably have the parameter-value-framework in mind and see the framework aspects of hadron as superfluous and would aim for a slim toolkit instead. But indeed, having some framework aspects makes it easier to use interactively.

I do not believe that one has to construct an additional container. The various members of the various mixins just need to be structured into a composition hierarchy. The `cf_orig` is then the container which contains the `cf_meta` and the actual data. The `cf_boot` is the container which contains the `cf_meta` and the bootstapped data; the original data would be lost after a `takeTimeDiff.cf` anyway.

All of these proposed changes are major refactorings of the whole library, so I guess it is mostly an academic exercise to discuss this. For the `raw_cf` class it should be still possible to change the design, although that will then be different to the current `cf`. Okay, I guess there is no obvious right answer to this …

Reply:
> I do not believe that one has to construct an additional container. The various members of the various mixins just need to be structured into a composition hierarchy. The cf_orig is then the container which contains the cf_meta and the actual data. The cf_boot is the container which contains the cf_meta and the bootstapped data; the original data would be lost after a takeTimeDiff.cf anyway.

I tend to agree with you more and more on this...